File: python/src/threads/agents/team.py
"""Team tools: the lead's log holds the shared task list and mailbox, and a
member changes it only through the lead's writer, so 's single writer makes each claim
atomic. A redelivered call never claims twice or sends twice: a claim the member already holds
and a message id already sent are answered without a second event."""

from pydantic import JsonValue
from pydantic.experimental.missing_sentinel import MISSING

from threads._generated.tools_v1 import (
    SendMessageInput,
    TeamTaskClaimInput,
    TeamTaskCreateInput,
    TeamTaskUpdateInput,
)
from threads.agents.scope import Scope
from threads.log import InjectedEvent, TeamMessageEvent, ToolCallData
from threads.loop.drafts import draft
from threads.loop.history import CallState
from threads.loop.results import As, result_draft
from threads.loop.runtime import Halt, Runtime, lost
from threads.result import Err
from threads.store import Draft


async def team_tool[D](scope: Scope[D], rt: Runtime, state: CallState) -> Halt | None:
    lead = scope.lead(rt)
    call = state.call.data
    event, text, failed = _decide(scope, lead, call)
    if event is not None and lead is not rt:
        # The lead's writer records the team state; this member's writer records the result.
        done = await lead.append(event)
        if isinstance(done, Err):
            return lost(done.error)
        event = None
    result = await result_draft(rt, call.call_id, text, As("executed", failed))
    done = await rt.append(*((event, result) if event is not None else (result,)))
    return lost(done.error) if isinstance(done, Err) else None


type Answer = tuple[Draft | None, str, bool]
"""The team event to append (if any), the result text, and whether it is an error."""


def _decide[D](scope: Scope[D], lead: Runtime, call: ToolCallData) -> Answer:
    """Ids are keyed to the call (`<member>/<call_id>`), so a call re-run after a crash finds
    what it already recorded and appends nothing (invariant 3)."""
    input = dict(call.input)
    member = scope.member()
    match call.name:
        case "team_task_create":
            args = TeamTaskCreateInput.model_validate(input)
            return _create(lead, f"{member}/{call.call_id}", args)
        case "team_task_claim":
            return _claim(lead, member, TeamTaskClaimInput.model_validate(input).task_id)
        case "team_task_update":
            return _update(lead, member, TeamTaskUpdateInput.model_validate(input))
        case _:
            args_m = SendMessageInput.model_validate(input)
            return _send(lead, member, f"{member}/{call.call_id}", args_m)


def _create(lead: Runtime, task_id: str, args: TeamTaskCreateInput) -> Answer:
    if task_id in lead.fold.tasks:
        return None, task_id, False
    blockers = [] if args.blocked_by is MISSING else [str(b) for b in args.blocked_by]
    unknown = [b for b in blockers if b not in lead.fold.tasks]
    if unknown:
        return None, f"unknown blockers: {', '.join(unknown)}", True
    data: dict[str, JsonValue] = {"task_id": task_id, "subject": args.subject}
    data["blocked_by"] = list[JsonValue](blockers)
    if args.description is not MISSING:
        data["description"] = args.description
    return draft("team_task_created", data), task_id, False


def _claim(lead: Runtime, member: str, task_id: str) -> Answer:
    tasks = lead.fold.tasks
    task = tasks.get(task_id)
    if task is not None and task.status == "claimed" and task.owner == member:
        return None, f"claimed {task_id}", False
    if task is None or task.status != "open":
        return None, f"can't claim: task {task_id} is not open", True
    blocked = [b for b in task.blocked_by if b not in tasks or tasks[b].status != "completed"]
    if blocked:
        return None, f"can't claim: task {task_id} has a blocker that is not completed", True
    claim = {"task_id": task_id, "member": member}
    return draft("team_task_claimed", claim), f"claimed {task_id}", False


def _update(lead: Runtime, member: str, args: TeamTaskUpdateInput) -> Answer:
    task = lead.fold.tasks.get(args.task_id)
    done = f"{args.task_id} {args.status}"
    if task is not None and task.owner == member and task.status == args.status:
        return None, done, False
    if task is None or task.status != "claimed" or task.owner != member:
        return None, f"{args.task_id} is not claimed by {member}", True
    data: dict[str, JsonValue] = {"task_id": args.task_id, "status": args.status}
    return draft("team_task_updated", data), done, False


def _send(lead: Runtime, member: str, message_id: str, args: SendMessageInput) -> Answer:
    sent = any(
        isinstance(e, TeamMessageEvent) and e.data.message_id == message_id for e in lead.events
    )
    if sent:
        return None, "sent", False
    data: dict[str, JsonValue] = {
        "message_id": message_id,
        "from": member,
        "to": args.to,
        "text": args.text,
    }
    return draft("team_message", data), "sent", False


async def deliver[D](scope: Scope[D], rt: Runtime) -> Halt | bool:
    """Messages to this member (or to everyone) it hasn't seen yet, as untrusted reference;
    delivery is deduplicated by message id, so a restart never injects one twice."""
    if scope.team is None and not scope.definition.subagents:
        return False
    member = scope.member()
    seen = {
        e.data.origin.id
        for e in rt.events
        if isinstance(e, InjectedEvent) and e.data.source == "agent"
    }
    drafts: list[Draft] = []
    for event in scope.lead(rt).events:
        if not isinstance(event, TeamMessageEvent):
            continue
        data = event.data
        addressed = data.to in (member, "*") and data.from_ != member
        if addressed and data.message_id not in seen and isinstance(data.text, str):
            note: dict[str, JsonValue] = {
                "source": "agent",
                "trust": "untrusted_reference",
                "origin": {"id": data.message_id},
                "text": f"{data.from_}: {data.text}",
            }
            drafts.append(draft("injected", note))
    if not drafts:
        return False
    done = await rt.append(*drafts)
    return lost(done.error) if isinstance(done, Err) else True

File: typescript/packages/core/src/loop/agents/team.ts
import type { EventOf } from "../../fold/state";
import {
  SendMessageInput,
  TeamTaskClaimInput,
  TeamTaskCreateInput,
  TeamTaskUpdateInput,
} from "../../tools/agent-inputs";
import { checkTask } from "../../validate/agents";
import { draft, TOOL } from "../drafts";
import type { Session } from "../session";
import type { Halt, Team } from "../types";

// Teams: the lead's log holds the task list and the mailbox, and members
// change them only through the lead's writer, so claims are atomic per the spec. Every act is
// keyed by the member's call, so a member re-dispatching a call after a restart gets the same
// answer and never claims or sends twice (F7.4).

type Call = EventOf<"tool_call">;
type Answer = ReturnType<Team["act"]>;

const ok = (output: string): Answer => ({ isError: false, output });
const no = (output: string): Answer => ({ isError: true, output });

/** The team whose lead is this session's thread. */
export function teamOf(s: Session): Team {
  return {
    act: (member, call) => act(s, member, call),
    inbox: (member) =>
      s.events.flatMap((e) =>
        e.type === "team_message" &&
        e.data.from !== member &&
        (e.data.to === member || e.data.to === "*")
          ? [e.data]
          : [],
      ),
  };
}

function act(s: Session, member: string, call: Call["data"]): Answer {
  const key = `${member}/${call.call_id}`;
  const { input } = call;
  switch (call.name) {
    case "team_task_create":
      return create(s, key, TeamTaskCreateInput.parse(input));
    case "team_task_claim":
      return claim(s, member, TeamTaskClaimInput.parse(input).task_id);
    case "team_task_update":
      return update(s, member, TeamTaskUpdateInput.parse(input));
    case "send_message":
      return send(s, member, key, SendMessageInput.parse(input));
    default:
      throw new Error(`${call.name} is not a team tool`);
  }
}

function appended(output: string, stopped?: Halt): Answer {
  return stopped === undefined ? ok(output) : no(stopped.message);
}

/** The task id is the creating call's key, so a repeated create is the same task. */
function create(
  s: Session,
  taskId: string,
  input: ReturnType<typeof TeamTaskCreateInput.parse>,
): Answer {
  if (s.fold.tasks.has(taskId)) return ok(taskId);
  const blockers = input.blocked_by ?? [];
  const unknown = blockers.find((id) => !s.fold.tasks.has(id));
  if (unknown !== undefined) return no(`no task ${unknown}`);
  const { subject, description } = input;
  return appended(
    taskId,
    s.append(
      draft.taskCreated({
        task_id: taskId,
        subject,
        blocked_by: blockers,
        ...(description === undefined ? {} : { description }),
      }),
    ),
  );
}

function claim(s: Session, member: string, taskId: string): Answer {
  const task = s.fold.tasks.get(taskId);
  if (task?.status === "claimed" && task.owner === member)
    return ok(`claimed ${taskId}`);
  const data = { task_id: taskId, member };
  const refused = checkTask(s.fold, { type: "team_task_claimed", data });
  if (refused !== undefined) return no(`can't claim: ${refused.message}`);
  return appended(`claimed ${taskId}`, s.append(draft.taskClaimed(data)));
}

function update(
  s: Session,
  member: string,
  input: ReturnType<typeof TeamTaskUpdateInput.parse>,
): Answer {
  const { task_id, status } = input;
  const task = s.fold.tasks.get(task_id);
  if (task?.status === status) return ok(`${task_id} ${status}`);
  if (task?.status !== "claimed" || task.owner !== member)
    return no(`${task_id} is not claimed by ${member}`);
  return appended(
    `${task_id} ${status}`,
    s.append(draft.taskUpdated({ task_id, status })),
  );
}

function send(
  s: Session,
  member: string,
  messageId: string,
  input: ReturnType<typeof SendMessageInput.parse>,
): Answer {
  if (s.fold.messageIds.has(messageId)) return ok("sent");
  return appended(
    "sent",
    s.append(
      draft.teamMessage({
        message_id: messageId,
        from: member,
        to: input.to,
        text: input.text,
      }),
    ),
  );
}

/** A team tool call in this thread: routed to its lead, or to itself as the lead. */
export function teamTool(s: Session, call: Call): Halt | undefined {
  const agents = s.config.agents;
  const answer =
    agents === undefined
      ? no("this agent has no team")
      : (agents.team ?? teamOf(s)).act(agents.name, call.data);
  return s.append(
    draft.toolResult(
      {
        call_id: call.data.call_id,
        is_error: answer.isError,
        origin: "executed",
        preview: answer.output,
      },
      TOOL,
    ),
  );
}

/**
 * Before a turn request: each team message for this agent not yet delivered here, as
 * untrusted reference keyed by its message_id, so a redelivery never injects it twice.
 */
export function deliverMessages(s: Session): Halt | undefined {
  const agents = s.config.agents;
  if (agents === undefined) return undefined;
  const seen = new Set(
    s.events.flatMap((e) =>
      e.type === "injected" && e.data.source === "agent"
        ? [e.data.origin.id]
        : [],
    ),
  );
  const fresh = (agents.team ?? teamOf(s))
    .inbox(agents.name)
    .filter((m) => !seen.has(m.message_id));
  if (fresh.length === 0) return undefined;
  return s.append(
    ...fresh.map((m) =>
      draft.injected({
        source: "agent",
        trust: "untrusted_reference",
        origin: { id: m.message_id },
        text: `${m.from}: ${m.text ?? ""}`,
      }),
    ),
  );
}
